problem-gson/src/main/java/org/zalando/problem/CustomProblemAdapter.java

```diff
@@ -0,0 +1,71 @@
+package org.zalando.problem;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.TypeAdapter;
+import com.google.gson.internal.Streams;
+import com.google.gson.reflect.TypeToken;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Arrays;
+import java.util.Optional;
+
+final class CustomProblemAdapter<T> extends TypeAdapter<T> {
+
+    private final boolean stackTraces;
+    private TypeAdapter<T> delegate;
+    private Gson gson;
+
+    CustomProblemAdapter(Gson gson, ProblemAdapterFactory parent, Class<T> type, boolean stackTraces) {
+        this.gson = gson;
+        this.stackTraces = stackTraces;
+        this.delegate = gson.getDelegateAdapter(parent, TypeToken.get(type));
+    }
+
+    @Override
+    public void write(JsonWriter out, T value) throws IOException {
+        JsonElement element = delegate.toJsonTree(value);
+        JsonObject jsonObject = element.getAsJsonObject();
+
+        // Handle `type` field.
+        URI problemType = Optional.ofNullable(jsonObject.remove("type"))
+                .map(TypeURIAdapter.INSTANCE::fromJsonTree).get();
+        jsonObject.add("type", TypeURIAdapter.INSTANCE.toJsonTree(problemType));
+
+        if (value instanceof AbstractThrowableProblem) {
+            // Flatten `parameters` field if AbstractThrowableProblem.
+            Optional.ofNullable(jsonObject.remove("parameters"))
+                    .map(JsonElement::getAsJsonObject)
+                    .ifPresent(params -> params.keySet().forEach(k -> jsonObject.add(k, params.get(k))));
+        }
+
+        if (value instanceof Throwable) {
+            // Get rid of unwanted fields.
+            jsonObject.remove("detailMessage");
+            jsonObject.remove("suppressedExceptions");
+            Optional.ofNullable(jsonObject.remove("stackTrace"))
+                    .filter(s -> stackTraces)
+                    .ifPresent(s -> {
+                        TypeAdapter<StackTraceElement[]> traceAdapter = gson.getAdapter(StackTraceElement[].class);
+                        StackTraceElement[] readTraces = traceAdapter.fromJsonTree(s);
```

Comment:
Couldn't we just use `value.getStackTrace()` here instead of re-parsing it?

Reply:
Oh yes 👍, missed that!